src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/BaseCodeActionService.cs

```diff
@@ -211,33 +215,47 @@ private IEnumerable<AvailableCodeAction> ConvertToAvailableCodeAction(IEnumerabl
-        // If same document has multiple codefixers (diagnostics with different fixers) will them be mapped as separate items.
-        protected async Task<ImmutableArray<DocumentWithFixProvidersAndMatchingDiagnostics>> GetDiagnosticsMappedWithFixAllProviders(FixAllScope scope, string fileName)
+        protected async Task<(DocumentId DocumentId, Diagnostic Diagnostic)> GetDocumentIdAndDiagnosticForGivenId(FixAllScope scope, Document document, string diagnosticId)
         {
-            ImmutableArray<DocumentDiagnostics> allDiagnostics = await GetCorrectDiagnosticsInScope(scope, fileName);
+            var allDocumentDiagnostics = await GetDiagnosticsAsync(scope, document);
 
-            var mappedProvidersWithDiagnostics = allDiagnostics
-                .SelectMany(diagnosticsInDocument =>
-                    DocumentWithFixProvidersAndMatchingDiagnostics.CreateWithMatchingProviders(codeFixesForProject.GetAllCodeFixesForProject(diagnosticsInDocument.ProjectId), diagnosticsInDocument));
+            foreach (var documentAndDiagnostics in allDocumentDiagnostics)
+            {
+                if (documentAndDiagnostics.Diagnostics.FirstOrDefault(d => d.Id == diagnosticId) is Diagnostic diagnostic)
+                {
+                    return (documentAndDiagnostics.Document.Id, diagnostic);
+                }
+            }
 
-            return mappedProvidersWithDiagnostics.ToImmutableArray();
+            return default;
+        }
+
+        protected ImmutableArray<CodeFixProvider> GetCodeFixProviders(Project project)
+        {
+            return _codeFixesForProject.GetAllCodeFixesForProject(project);
+        }
+
+        protected CodeFixProvider? GetCodeFixProviderForId(Document document, string id)
+        {
+            // If Roslyn ever comes up with a UI for selecting what provider the user prefers, we might consider replicating.
+            // https://github.com/dotnet/roslyn/issues/27066
+            return _codeFixesForProject.GetAllCodeFixesForProject(document.Project).FirstOrDefault(provider => provider.HasFixForId(id));
         }
 
-        private async Task<ImmutableArray<DocumentDiagnostics>> GetCorrectDiagnosticsInScope(FixAllScope scope, string fileName)
+        protected async Task<ImmutableArray<DocumentDiagnostics>> GetDiagnosticsAsync(FixAllScope scope, Document document)
         {
             switch (scope)
             {
                 case FixAllScope.Solution:
-                    var documentsInSolution = Workspace.CurrentSolution.Projects.SelectMany(x => x.Documents).Select(x => x.FilePath).ToImmutableArray();
-                    return await diagnostics.GetDiagnostics(documentsInSolution);
+                    var documentsInSolution = document.Project.Solution.Projects.SelectMany(p => p.Documents).ToImmutableArray();
+                    return await _diagnostics.GetDiagnostics(documentsInSolution, skipCache: true);
```

Comment:
what are the advantages / disadvantages of using `skipCache` here?

Reply:
reread the pr, I'm slow ignore me here, lol.

Reply:
Since I'm moving the caching layer to use a CWT, I can remove the `skipCache` parameter entirely now.